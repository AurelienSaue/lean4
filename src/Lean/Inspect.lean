/-
Copyright (c) 2021 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Daniel Selsam
-/
import Lean.Compiler.IR

namespace Lean.Inspect

inductive Object : Type
  | ctor        : Nat → Array Object → Object
  | closure     : (fileName symbolName : Option String) → Nat → Array Object → Object
  | scalar      : Nat → Object
  | unsupported : Object /- TODO(dselsam): other kinds -/
  deriving Repr, Inhabited, BEq

end Inspect

@[extern "lean_inspect"]
constant inspect (thing : PNonScalar) : IO Inspect.Object

end Lean
