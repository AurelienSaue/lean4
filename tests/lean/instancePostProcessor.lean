import Lean.Meta

class Zero (α : Type u) := (zero : α)
class One  (α : Type u) := (one : α)

universes u
variable {α : Type u} [Zero α] [One α] [Add α]

def bit0 (x : α) : α := x + x
def bit1 (x : α) : α := bit0 x + One.one

instance instOne2Inhabited : Inhabited α := ⟨Zero.zero⟩

open Lean Lean.Meta

partial def nat2bits (n : Nat) : α :=
  if n == 0 then Zero.zero
  else if n == 1 then One.one
  else if n % 2 == 1 then bit1 (nat2bits (n / 2))
  else bit0 (nat2bits (n / 2))

instance instBits2Nat (n : Nat) : OfNat α n := ⟨nat2bits n⟩

open Lean.Meta (mkAppM instantiateMVars)

partial def nat2bitsExpr (lvl : Level) (type zero one add : Expr) (n : Nat) : MetaM Expr := do
  if      n == 0 then mkAppN (mkConst `Zero.zero [lvl]) #[type, zero]
  else if n == 1 then mkAppN (mkConst `One.one [lvl]) #[type, one]
  else
    let rec ← nat2bitsExpr lvl type zero one add (n/2)
    if n % 2 == 0 then mkAppN (mkConst `bit0 [lvl]) #[type, add, rec]
    else               mkAppN (mkConst `bit1 [lvl]) #[type, one, add, rec]

@[instancePostProcessor instBits2Nat]
def instBits2NatPostProcessor : Lean.Meta.InstancePostProcessor := {
  apply := λ e => do
    let lvl  := e.getAppFn.constLevels!.head!
    let #[type, zero, one, add, n] ← pure e.getAppArgs | throwError "foo"
    match (← instantiateMVars n).natLit? with
    | none   => pure none
    | some n =>
      let x ← nat2bitsExpr lvl type zero one add n
      pure $ some $ mkAppN (mkConst `OfNat.mk [lvl]) #[type, mkNatLit n, x]
}

set_option pp.all true
#check (0 : α)
#check (1 : α)
#check (2 : α)
#check (5 : α)
