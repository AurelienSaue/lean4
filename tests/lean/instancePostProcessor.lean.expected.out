@OfNat.ofNat.{u} α 0 (@OfNat.mk.{u} α 0 (@Zero.zero.{u} α inst✝²)) : α
@OfNat.ofNat.{u} α 1 (@OfNat.mk.{u} α 1 (@One.one.{u} α inst✝¹)) : α
@OfNat.ofNat.{u} α 2 (@OfNat.mk.{u} α 2 (@bit0.{u} α inst✝ (@One.one.{u} α inst✝¹))) : α
@OfNat.ofNat.{u} α 5 (@OfNat.mk.{u} α 5 (@bit1.{u} α inst✝¹ inst✝ (@bit0.{u} α inst✝ (@One.one.{u} α inst✝¹)))) : α
